pkg/insight/insightmetrics/metrics.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package insightmetrics
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"github.com/pipe-cd/pipe/pkg/insight/insightstore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+const (
+	projectKey = "project"
+	appKindKey = "app_kind"
+)
+
+type insightMetricsCollector struct {
+	insightStore insightstore.Store
+	projectStore datastore.ProjectStore
+
+	applicationDesc *prometheus.Desc
+}
+
+func NewInsightMetricsCollector(is insightstore.Store, ps datastore.ProjectStore) prometheus.Collector {
+	return &insightMetricsCollector{
+		insightStore: is,
+		projectStore: ps,
+		applicationDesc: prometheus.NewDesc(
+			"insight_application_total",
+			"Number of applications currently controlled by control plane",
+			[]string{projectKey, appKindKey},
+			nil,
+		),
+	}
+}
+
+func (i *insightMetricsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- i.applicationDesc
+}
+
+func (i *insightMetricsCollector) Collect(ch chan<- prometheus.Metric) {
+	applicationCounts, err := i.collectApplicationCount()
```

Comment:
To be honest I have concerns about the frequency of this execution. This will be executed every time when a Prometheus instance comes to scrape the metrics. I prefer to store data in memory to return it immediately. But let's apply this way this time to see how it works under our current conditions and improve it later if needed.

Reply:
Just feel the same, lets see how it works in devs for now 👍